sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala

```diff
@@ -287,4 +293,63 @@ class UserDefinedTypeSuite extends QueryTest with SharedSparkSession with Parque
     checkAnswer(spark.createDataFrame(data, schema).selectExpr("typeof(a)"),
       Seq(Row("array<double>")))
   }
+
+  test("Allow merge UserDefinedType into a native DataType") {
```

Comment:
How about the simplified one like this?
```
  test("Allow merge UserDefinedType into a native DataType") {
    // Register the UDT
    UDTRegistration.register(
      classOf[XMLGregorianCalendar].getName,
      classOf[MyXMLGregorianCalendarUDT].getName)

    withTempDir { dir =>
      val gregorianCalendar = new GregorianCalendar(1925, 5, 20, 19, 25)
      // Equivalent of above (the year minus 1900)
      val timestamp = new Timestamp(25, 5, 20, 19, 25, 0, 0)

      val data = Seq(Row(new XMLGregorianCalendarImpl(gregorianCalendar)))
      val rdd = spark.sparkContext.parallelize(data)
      val schema = StructType(StructField("dt", new MyXMLGregorianCalendarUDT) :: Nil)
      val df = spark.sqlContext.createDataFrame(rdd, schema)

      df.write.mode(SaveMode.Append).save(dir.getCanonicalPath)
      // We should be able to write a second time, and Spark should be able to resolve the types
      df.write.mode(SaveMode.Append).save(dir.getCanonicalPath)

      val records = spark.read.load(dir.getCanonicalPath)
      records.printSchema()
      checkAnswer(records, Row(timestamp) :: Row(timestamp) :: Nil)
    }
  }
```
But, it seems this current test is not related to this fix (I mean this test passed without the fix). Can you check again?

Reply:
Perfect, much cleaner. Thanks for suggesting.

I'm pretty sure that it will fail without the fix. It will fail when it will write the second time. Since we've saved the `XMLGregorianCalendarImpl` as a timestamp (this is why we've overwritten the `jvalue`), Spark will recognize it as a regular timestamp. At then next append we're trying to merge an XMLGregorianCalendar into a timestamp, which isn't allowed with the additional rule.

Reply:
Just tried to copy this test only into the current master again, but it passed. I think the other reviewers need to double-check.

Reply:
Ok, let's remove the rule, and I'll retrigger the test

Reply:
You're right, let me investigate.

Reply:
Interesting. When appending to the table, Spark does not check compatibility on write, when using Delta this is the case.

Reply:
Ok, I've rewritten the test, and not it fails on current master:
```
scala> val df = gregorianCalendarDf.union(timestampDf)
org.apache.spark.sql.AnalysisException: Union can only be performed on tables with the compatible column types. timestamp <> timestamp at the first column of the second table;;
'Union
:- LogicalRDD [dt#11], false
+- LogicalRDD [dt#14], false

  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis$class.failAnalysis(CheckAnalysis.scala:43)
  at org.apache.spark.sql.catalyst.analysis.Analyzer.failAnalysis(Analyzer.scala:95)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis$$anonfun$checkAnalysis$1$$anonfun$apply$12$$anonfun$apply$13.apply(CheckAnalysis.scala:294)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis$$anonfun$checkAnalysis$1$$anonfun$apply$12$$anonfun$apply$13.apply(CheckAnalysis.scala:291)
  at scala.collection.immutable.List.foreach(List.scala:392)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis$$anonfun$checkAnalysis$1$$anonfun$apply$12.apply(CheckAnalysis.scala:291)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis$$anonfun$checkAnalysis$1$$anonfun$apply$12.apply(CheckAnalysis.scala:280)
  at scala.collection.immutable.List.foreach(List.scala:392)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis$$anonfun$checkAnalysis$1.apply(CheckAnalysis.scala:280)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis$$anonfun$checkAnalysis$1.apply(CheckAnalysis.scala:86)
  at org.apache.spark.sql.catalyst.trees.TreeNode.foreachUp(TreeNode.scala:127)
  at org.apache.spark.sql.catalyst.analysis.CheckAnalysis$class.checkAnalysis(CheckAnalysis.scala:86)
  at org.apache.spark.sql.catalyst.analysis.Analyzer.checkAnalysis(Analyzer.scala:95)
  at org.apache.spark.sql.catalyst.analysis.Analyzer$$anonfun$executeAndCheck$1.apply(Analyzer.scala:108)
  at org.apache.spark.sql.catalyst.analysis.Analyzer$$anonfun$executeAndCheck$1.apply(Analyzer.scala:105)
  at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$.markInAnalyzer(AnalysisHelper.scala:201)
  at org.apache.spark.sql.catalyst.analysis.Analyzer.executeAndCheck(Analyzer.scala:105)
  at org.apache.spark.sql.execution.QueryExecution.analyzed$lzycompute(QueryExecution.scala:57)
  at org.apache.spark.sql.execution.QueryExecution.analyzed(QueryExecution.scala:55)
  at org.apache.spark.sql.execution.QueryExecution.assertAnalyzed(QueryExecution.scala:47)
  at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:78)
  at org.apache.spark.sql.Dataset.withSetOperator(Dataset.scala:3424)
  at org.apache.spark.sql.Dataset.union(Dataset.scala:1862)
  ... 40 elided
```
I'm creating two DF's, both representing time, and one of them a UDT. I'm trying to merge them. This is very similar to what I'm doing with Delta. With Delta, using `SaveMode.Overwrite` it will still check the compatibility with the existing schema.

Reply:
@maropu I'm unable to reproduce the issue without Delta. Since plan Spark does not check schema compatibility on write, I'm unable to implement an end to end integration test. I tried to fix this by taking the union of the two DF's, but this took another branch in the codebase, and it failed on `TypeCoercion`.

Please advise.

Reply:
After a good night of sleep, I've came up with an integration test, that will mimic the behavior as in Delta.